File: java/triangle/src/main/java/Triangle.java
class Triangle {

    final private double side1, side2, side3;

    Triangle(double side1, double side2, double side3) throws TriangleException {
        if(Double.compare(side1+side2, side3) <= 0 ||
           Double.compare(side1+side3, side2) <= 0 ||
           Double.compare(side3+side2, side1) <= 0)
            throw new TriangleException();

        this.side1 = side1;
        this.side2 = side2;
        this.side3 = side3;
    }

    boolean isEquilateral() {
        return Double.compare(side1, side2) == 0 && Double.compare(side3, side2) == 0;
    }

    boolean isIsosceles() {
        return Double.compare(side1, side2) == 0 || Double.compare(side3, side1) == 0 || Double.compare(side3, side2) == 0;
    }

    boolean isScalene() {
        return !isIsosceles();
    }

}

File: csharp/triangle/Triangle.cs
using System;

public static class Triangle
{
    public static bool IsScalene(double a, double b, double c)
        => IsTriangle(a, b, c) && !IsIsosceles(a, b, c);

    public static bool IsIsosceles(double a, double b, double c)
        => IsTriangle(a, b, c) && (a == b || a == c || b == c);

    public static bool IsEquilateral(double a, double b, double c)
        => ZeroCheck(a, b, c) && a == b && b == c;

    private static bool ZeroCheck(double a, double b, double c)
        => a != 0d && b != 0d && c != 0d; 

    private static bool IsTriangle(double a, double b, double c)
        => a + b >= c
        && a + c >= b
        && b + c >= a;
}